File: Algorithms/Sorting Algorithms/Insertion_Sort.cpp
#include <iostream>

using namespace std;

/**
 * @brief Generic insertion sort algorithm
 * 
 * @tparam T (generic type)
 * @param arr (array of type T)
 * @param size (array size)
 */
template <class T>
void insertion_sort(T arr[], int size) 
{
	for (size_t i = 1; i < size; i++)
	{
		int j = i - 1;
		while (j >= 0 && arr[j] > arr[j+1]) 
		{
			swap(arr[j], arr[j+1]);
			j -= 1;
		}
	}
}

int main() {
	int test[4] = {1, 4, 3, 2};
	insertion_sort(test, 4);

	for (size_t i = 0; i < 4; i++)
	{
		cout << test[i] << endl;
	}

	return 0;
}

File: src/main/java/Algorithms/SortingAlgorithms/InsertionSort.java
package Algorithms.SortingAlgorithms;

public class InsertionSort {
	/*
	 * Swaps the elements at the given indices
	 */
	private static <T> void swap(T[] array, int index1, int index2) {
		T hold = array[index1];
		array[index1] = array[index2];
		array[index2] = hold;
	}

	/**
	 * Insertion sort algorithm
	 * @param <T> generic type
	 * @param arr array to be sorted
	 * @return sorted array
	 */
	public <T extends Comparable<T>> T[] sort(T[] arr) {
		for (int i = 1; i < arr.length; i++) {
			int j = i - 1;	
			while (j >= 0 && arr[j].compareTo(arr[j + 1]) > 0) {
				swap(arr, j, j + 1);	
				j -= 1;
			}
		}

		return arr;
	}
}
